cpp/src/arrow/util/meson.build

```diff
@@ -92,3 +92,242 @@ configure_file(
+        'string_builder.h',
+        'string.h',
+        'task_group.h',
+        'tdigest.h',
+        'test_common.h',
+        'thread_pool.h',
+        'time.h',
+        'tracing.h',
+        'trie.h',
+        'type_fwd.h',
+        'type_traits.h',
+        'ubsan.h',
+        'union_util.h',
+        'unreachable.h',
+        'uri.h',
+        'utf8.h',
+        'value_parsing.h',
+        'vector.h',
+        'visibility.h',
+        'windows_compatibility.h',
+        'windows_fixup.h',
+    ],
+    subdir: 'arrow/util',
+)
+
+utility_test_srcs = [
+    'align_util_test.cc',
+    'atfork_test.cc',
+    'byte_size_test.cc',
+    'byte_stream_split_test.cc',
+    'cache_test.cc',
+    'checked_cast_test.cc',
+    'compression_test.cc',
+    'decimal_test.cc',
+    'float16_test.cc',
+    'fixed_width_test.cc',
+    'formatting_util_test.cc',
+    'key_value_metadata_test.cc',
+    'hashing_test.cc',
+    'int_util_test.cc',
+    'io_util_test.cc',
+    'iterator_test.cc',
+    'list_util_test.cc',
+    'logger_test.cc',
+    'logging_test.cc',
+    'math_test.cc',
+    'queue_test.cc',
+    'range_test.cc',
+    'ree_util_test.cc',
+    'reflection_test.cc',
+    'rows_to_batches_test.cc',
+    'small_vector_test.cc',
+    'span_test.cc',
+    'stl_util_test.cc',
+    'string_test.cc',
+    'tdigest_test.cc',
+    'test_common.cc',
+    'time_test.cc',
+    'tracing_test.cc',
+    'trie_test.cc',
+    'uri_test.cc',
+    'utf8_util_test.cc',
+    'value_parsing_test.cc',
+]
+
+if host_machine.system() == 'windows'
+    # This manifest enables long file paths on Windows 10+
+    # See https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#enable-long-paths-in-windows-10-version-1607-and-later
+    if cpp_compiler.get_id() == 'msvc'
+        utility_test_sources += ['io_util_test.manifest']
+    else
+        utility_test_sources += ['io_util_test.rc']
+    endif
+endif
+
+exc = executable(
+    'arrow-utility-test',
+    sources: utility_test_srcs,
+    dependencies: [arrow_dep, filesystem_dep, gtest_dep, gmock_dep],
+    link_with: [arrow_test_lib],
+    implicit_include_directories: false,
+)
+test('arrow-utility-test', exc)
+
+util_tests = {
+    'arrow-async-utility-test': {
+        'sources': [
+            'async_generator_test.cc',
+            'async_util_test.cc',
+            'test_common.cc',
+        ],
+    },
+    'arrow-bit-utility-test': {
+        'sources': [
+            'bit_block_counter_test.cc',
+            'bit_util_test.cc',
+            'rle_encoding_test.cc',
+        ],
+    },
+    'arrow-threading-utility-test': {
+        'sources': [
+            'cancel_test.cc',
+            'counting_semaphore_test.cc',
+            'future_test.cc',
+            'task_group_test.cc',
+            'test_common.cc',
+            'thread_pool_test.cc',
+        ],
+    },
+    'arrow-crc32-test': {
+        'sources': ['crc32_test.cc'],
+        'dependencies': [filesystem_dep],
+    },
+}
+
+if needs_tests
```

Comment:
Do we need this?

Reply:
We do. I think there are ways to improve the terminology, but technically the `arrow_test_dep` becomes a disabler only if `needs_testing` is false. It is possible for `needs_tests` to be false while `needs_testing` is true. Without this extra condition these tests would build when `-Dtests=false` and `-Dtesting=true`

I do wonder if it wouldn't be better to have an `arrow_test_dep` and separately an `arrow_testing_dep`

Reply:
If we need this here, we need it for `test('arrow-utility-test', exc)` too?

It seems that we don't need both `arrow_test_dep` and `arrow_testing_dep`. Can we use this instead?

```meson
if needs_tests
    arrow_test_dep = declare_dependency(
        link_with: [arrow_test_lib],
        dependencies: [arrow_dep, filesystem_dep, gmock_dep, gtest_main_dep],
    )
else
    arrow_test_dep = disabler()
endif
```

Reply:
Nice catch on the arrow-utility-test - I can do that.

With respect to the `if needs_tests`, it depends how much we want to adhere to what CMake has in place. In fact, we used to use that same condition, but I had to change it from `if needs_tests` to `if needs_testing` in the benchmarks PR. 

https://github.com/apache/arrow/pull/45793/files

Unless I am misreading the CMake configuration (which is certainly possible) it seems like it is possible to disable tests but turn the benchmarks on. In such a case, you would still need the current `arrow_test_dep` dependency to build benchmarks; the change you are proposing would add a dependency between the test suite and benchmark suite

Reply:
Can we use separated dependency for benchmarks something like the following?

```meson
if needs_tests
    arrow_test_dep = declare_dependency(
        link_with: [arrow_test_lib],
        dependencies: [arrow_dep, filesystem_dep, gmock_dep, gtest_main_dep],
    )
else
    arrow_test_dep = disabler()
endif

if needs_benchmarks
  arrow_benchmark_dep = declare_dependency(
        dependencies: [benchmark_dep, arrow_test_dep],
    )
else
  arrow_benchmark_dep = disabler()
end
```

Reply:
Yes we can. Nice idea!